device/transport/mqtt/src/mqtt_twin_receiver.ts

```diff
@@ -47,80 +50,136 @@ export class MqttTwinReceiver extends EventEmitter {
+            /* Codes_SRS_NODE_DEVICE_MQTT_TWIN_RECEIVER_18_024: [** When the `error` event is emitted, the first parameter shall be an error object obtained via the MQTT `translateError` module. **]** */
+            this.emit(MqttTwinReceiver.errorEvent, translateError(err));
+          }
         });
       }
-    } else if (eventName === MqttTwinReceiver.postEvent) {
-      if (EventEmitter.listenerCount(this, MqttTwinReceiver.postEvent) === 0) {// array of listeners gets updated _after_ firing this event
-        /* Codes_SRS_NODE_DEVICE_MQTT_TWIN_RECEIVER_18_018: [** When a listener is added to the post event, the appropriate topic shall be asynchronously subscribed to. **]** */
-        process.nextTick(() => {
-          self._mqtt.subscribe(postTopic, { qos: 0 }, (err, transportObject) => {
+    });
+
+    this.on('removeListener', (eventName) => {
+      if (eventName === MqttTwinReceiver.responseEvent || eventName === MqttTwinReceiver.postEvent) {
+        if (this.listeners('response').length + this.listeners('post').length <= 0) {
+          this._fsm.handle('unsubscribe', (err) => {
             if (err) {
-              self._handleError(err);
-            } else {
-              /* Codes_SRS_NODE_DEVICE_MQTT_TWIN_RECEIVER_18_025: [** If the `subscribed` event is subscribed to, a `subscribed` event shall be emitted after an MQTT topic is subscribed to. **]** */
-              /* Codes_SRS_NODE_DEVICE_MQTT_TWIN_RECEIVER_18_028: [** When the `subscribed` event is emitted, the parameter shall be an object which contains an `eventName` field and an `transportObject` field. **]** */
-              /* Codes_SRS_NODE_DEVICE_MQTT_TWIN_RECEIVER_18_027: [** When the `subscribed` event is emitted because the post MQTT topic was subscribed, the `eventName` field shall be the string 'post' **]** */
-              /* Codes_SRS_NODE_DEVICE_MQTT_TWIN_RECEIVER_18_029: [** When the subscribed event is emitted, the `transportObject` field shall contain the object returned by the library in the subscription response. **]** */
-              self.emit(MqttTwinReceiver.subscribedEvent, { 'eventName' : MqttTwinReceiver.postEvent, 'transportObject' : transportObject });
+              debug('error while unsubscribing: ' + err.toString());
+              /* Codes_SRS_NODE_DEVICE_MQTT_TWIN_RECEIVER_18_024: [** When the `error` event is emitted, the first parameter shall be an error object obtained via the MQTT `translateError` module. **]** */
+              this.emit(MqttTwinReceiver.errorEvent, translateError(err));
             }
           });
-        });
+        }
       }
-    }
+    });
+
+    const messageHandler = this._onMqttMessage.bind(this);
+
+    this._fsm = new machina.Fsm({
+      namespace: 'device-mqtt',
+      initialState: 'unsubscribed',
+      states: {
+        unsubscribed: {
+          _onEnter: (callback, err) => {
+            this._mqtt.removeListener('message', messageHandler);
```

Comment:
>this._mqtt.removeListener('message', messageHandler); [](start = 12, length = 53)

I am so annoyed that a 'message' event could be an MQTT packet that is emitted by a lower layer or it could be a C2D message that is emitted at this layer.  I doubt if we want to change this right now, but AAARGH.

Reply:
take a deep breath. it's all internal implementation details. FWIW I agree - 'common' message translation should happen only in the highest layer.

---
In reply to: [145828101](https://github.com/Azure/azure-iot-sdk-node/pull/111#discussion_r145828101) [](ancestors = 145828101)